### ISSUE TITLE
Change getAll pings to run only once instead of for each request - this Improves display speed

### ISSUE THREAD

**Issue by RGA29** [oslabs-beta/LatenSee#68]

**Comment by EvanCG** [comment]
Started a benchmark with current state after building testing for this portion of the app and recording the results here. For these tests, I've locked the size of `data.csv` at 212,778 records.

When loading the page a few times, I found that the data fetch took an average of 3.05s to complete. 47.2% of that time is waiting for the server, which is the main place to look for improvement.

| #       | Waiting For Server Response | Content Download | Total |
|---------|-----------------------------|------------------|-------|
| 1       | 1.41s                       | 1.59s            | 3.00s |
| 2       | 1.47s                       | 1.61s            | 3.08s |
| 3       | 1.44s                       | 1.65s            | 3.09s |
| 4       | 1.45s                       | 1.60s            | 3.05s |
| 5       | 1.44s                       | 1.61s            | 3.05s |
| Average | 1.44s                       | 1.61s            | 3.05s |

This implies we can get savings in two ways:
- We can reduce the time it takes for the server to process results (the primary purpose of this issues)
- We can reduce the amount of content that is transferred to the front end (potentially opportunity for a future refactor).

A `GET` reqeuest to the `/api/data` endpoint runs through two pieces of middleware - `dataController.getData` and `dataController.getRuns`. This refactor will focus on reducing the second of those, and the processing time to manage it.

Running a fetch request for this endpoint on the server yields an average completion time of **973ms** after 20 requests, so this will be the 'before' benchmark'.

_datacontrollerTiming.test.js_:
```
Timing results is:  [
       993,  959, 1004, 944, 1007,
       955,  979,  942, 974,  961,
      1022,  937, 1012, 960,  985,
       937, 1000,  961, 983,  945
]
```

**Comment by EvanCG** [comment]
Updated benchmarking, and page load of the `data` fetch has improved **-15.2%** overall, while the portion of waiting for a server response has improved about **-30.4%**. When using the testing timing folder, the server-side run time of the fetch has improved -44.5% to **539.6ms**


| #           | Waiting For Server Response | Content Download | Total  |
|-------------|-----------------------------|------------------|--------|
| 1           | 1.00s                       | 1.59s            | 2.59s  |
| 2           | 1.00s                       | 1.57s            | 2.57s  |
| 3           | 1.00s                       | 1.59s            | 2.59s  |
| 4           | 1.00s                       | 1.60s            | 2.60s  |
| 5           | 1.02s                       | 1.58s            | 2.60s  |
| Average     | 1.00s                       | 1.59s            | 2.59s  |
| Improvement | -30.4%                      | -1.6%            | -15.2% |

_datacontrollerTiming.test.js_
```
    Timing results is:  [
      548, 546, 550, 540, 550,
      524, 547, 526, 553, 518,
      554, 525, 553, 522, 555,
      524, 555, 523, 557, 522
    ]
```

Continued testing shows that most of the time in this function is now taken up by retrieving the full set of data from grabbing all of the rows of data. For example, see this timing:
```
Finished run, data is:  {
  fullRunInterval: 296, // total time to complete getRuns
  grabAllRowsInterval: 272, // Time to grab all data
  builtAggregatorInterval: 0,
  processedRecordsInterval: 24, // time to calculate averages etc
  createdResultsInterval: 0
}
```

Next steps for optimization are probably to explore how to more quickly read files from the CSV, and potentially splitting the need to send all of the raw data back to the front-end (to save download time).

**Comment by EvanCG** [comment]
Made a quick change that largely truncated the data being sent back in `res.locals.all` which nearly completely eliminated the additional step of downloading data. This resulted in additional gains on the client side, and the server-side fetch timing also seems to have improved as a result:

| #                       | Waiting For Server Response | Content Download | Total  |
|-------------------------|-----------------------------|------------------|--------|
| 1                       | 0.92s                       | 0.00s            | 0.92s  |
| 2                       | 0.98s                       | 0.00s            | 0.98s  |
| 3                       | 0.97s                       | 0.00s            | 0.97s  |
| 4                       | 0.98s                       | 0.00s            | 0.98s  |
| 5                       | 0.96s                       | 0.00s            | 0.97s  |
| **Average**                 | **0.96s**                       | **0.00s**            | **0.96s**  |
| _Improvement vs Baseline_ | _-33.3%_                      | _-100.0%_          | _-68.5%_ |

_dataControllerTiming.test.js_:
```
    Timing results is:  [
      304, 303, 304, 297, 299,
      302, 307, 298, 299, 306,
      297, 298, 306, 296, 298,
      308, 299, 299, 308, 297
    ]
```

Average length of the timing test is now **301.25ms**

**Comment by EvanCG** [comment]
Built out a test suite (`csvReadingTiming.test.js`) to try and benchmark to minimum time needed to read the entirety of the csv file using the `csv-parser` package.

Consistently found the average csv parsing time to be 298ms and getRuns to be 311.1ms. This implies that 95.8% of the latency introduced by this function is due to the required reading of the CSV file to generate stats and a raw read rate of ~1.4ms per 1k lines.

This rate is materially faster than the benchmark rates shared in [csv-parser](https://www.npmjs.com/package/csv-parser) documentation which leads me to conclude there isn't much additional benefit from reducing this without refactoring the implementation of reading the file itself (i.e. to break after recognizing that rows are outside of the date range in question).

| Source          | This Repo                | NPM Example       | NPM Example            |
|-----------------|--------------------------|-------------------|------------------------|
| File            | _csvReadingTiming.test.js_ | _large-dataset.csv_ | _option-maxRowBytes.csv_ |
| Read time              | 298ms                    | 73ms              | 39ms                   |
| Lines   in CSV  |        212,778           | 7,268             | 4,577                  |
| **ms /   1k lines** | **1.40ms**                   | **10.04ms**           | **8.52ms**                 |

For now, we've attained a -68.5% decrease in overall client load time, which we can update.

**Comment by EvanCG** [comment]
Resolved with #72 .
